### PR TITLE
create "what's new in C# 7.3"

### LINKED ISSUE #5289: Write the "What's new in C# 7.3" topic
Each individual language feature is part of the [C# 7.3 / VB15.7] project. That project also contains a couple catch-up items. 

All these new features need to be put together in one topic for readers to see all the new features.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #5289

This PR adds the overview to "What's new in C# 7.3". I updated the wording in the TOC to remove the repeated "What's new" phrasing.

/cc @jcouv @madstorgersen

/cc @Welchen

This PR relies on new files added in PR #5408

## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

### CODE REVIEW

File: docs/csharp/whats-new/csharp-7-3.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+- You can reassign `ref` local variables.
+- You can use initializers on `stackalloc` arrays.
+- You can use `fixed` statements with any type that supports a pattern.
+- You can use additional generic constraints.
+
+The following enhancements were made to existing features:
+
+- You can test `==` and `!=` with tuple types.
+- You can use expression variables in more locations.
+- You may attach attributes to the backing field of auto implemented properties.
```
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-3.md]: Nit: auto-implemented

File: docs/csharp/whats-new/csharp-7-3.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+- Method resolution when arguments differ by `in` has been improved.
+- Overload resolution new has fewer ambiguous cases.
```
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-3.md]: Typo: new -> now

File: docs/csharp/whats-new/csharp-7-3.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+The new compiler options are:
+
+- `deterministic` to ensure that subsequent compilations of the same source generate the same binary output.
```
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-3.md]: I'm surprised. "deterministic" is not a new option. It was there for a while.

Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: Well, we got behind. I can move it to the correct version. Do you know when it was first added?

**Review by jcouv — APPROVED** [review]
LGTM Thanks!

File: docs/csharp/whats-new/csharp-7-3.md (hunk 3 above)
Comment (jcouv) [reply docs/csharp/whats-new/csharp-7-3.md]: `-deterministic` was added in C# 6.0.

File: docs/csharp/whats-new/csharp-7-3.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaakov-h) [reply docs/csharp/whats-new/csharp-7-3.md]: “now”?

**Comment by MihaMarkic** [comment]
LGTM as well. Perhaps you could add a couple of enhancements:
- add a _Span<T> arr = stackalloc [] {1, 2, 3};_  sample
- add either Enum or Delegate or both samples to generic constraints

**Comment by MihaMarkic** [comment]
OTOH, is 

> When the in argument modifier was added, these two methods would cause an ambiguity:
> 
> static void M(S arg);
> static void M(in S arg);

Is this real? I've tested it with 7.2 and it works:

```
        static void X(int a) {}
        static void X(in int a) {}
```

File: docs/csharp/whats-new/csharp-7-3.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+There are two main themes to the C# 7.3 release. One theme provides features that enable safe code to be as performant as unsafe code. The second theme provides incremental improvements to existing features. In addition, new compilers were added in this release.
+
+The following new features support the theme of better performance for safe code:
+
+- You can access fixed fields without pinning.
+- You can reassign `ref` local variables.
+- You can use initializers on `stackalloc` arrays.
+- You can use `fixed` statements with any type that supports a pattern.
+- You can use additional generic constraints.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: It could go in either section. I put it here because of the `unmanaged` constraint. That enables some more perf oriented generic programming.

File: docs/csharp/whats-new/csharp-7-3.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: "overload resolution now..."

**Comment by BillWagner** [comment]
> Perhaps you could add a couple of enhancements:
>add a Span<int> arr = stackalloc [] {1, 2, 3}; sample
I like this idea. I added it.
> add either Enum or Delegate or both samples to generic constraints
If want to keep this article reasonably short. The articles linked at the end of that section have samples of both Enum and Delegate, in addition to unmanaged.

**Comment by BillWagner** [comment]
> Is this real? I've tested it with 7.2 and it works:

Yes, and I added a note that this is a bug fix, and the new resolution applies even if C# 7.2 is selected as the language version.

File: docs/csharp/whats-new/csharp-7-3.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (pkulikov) [reply docs/csharp/whats-new/csharp-7-3.md]: OK, I see, thank you for the reply.

File: docs/csharp/whats-new/index.md [hunk 5]
```diff
@@ -12,52 +12,38 @@ the C# language. The following links provide detailed information on the
...
-
+> The C# language relies on types and methods in a *standard library* for some of the features. One example is exception processing. Every `throw` statement or expression is checked to ensure the object being thrown is derived from <xref:System.Exception>. Similarly, every `catch` is checked to ensure that the type being caught is derived from <xref:System.Exception>. Each version may add new requirements. To use the latest language features in older environments, you may need to install specific libraries. These dependencies are documented in the page for each specific version. You can learn more about the [relationships between language and library](relationships-between-language-and-library.md) for background on this dependency.
 
+* [C# 7.3](csharp-7-3.md):
+  - This page describes the latest features in the C# language. C# 7.3 is currently available in [Visual Studio 2017 version 15.7](https://www.visualstudio.com/vs/whatsnew/), and in the [.NET Core 2.1 SDK Preview](../../core/whats-new/index.md).
```
Comment (mairaw) [inline docs/csharp/whats-new/index.md]: 2.1.300 RC1?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+### Public or OSS signing
+
+The `-publicsign` compiler option instructs the compiler to sign the assembly using a public key. The assembly is marked as signed, but the signature is taken from the public key. This option enables you to build signed assemblies from open-source projects using a public key.
+
+For more information, see the [-publicsign compiler option](../language-reference/compiler-options/publicsign-compiler-option.md) article.
+
+### pathmap
+
+The `-pathmap` compiler option instructs the compiler to replace source paths from the build environment with mapped source paths. The pathmap option controls the source path written by the compiler to PDB files or for the <xref:System.Runtime.CompilerServices.CallerFilePathAttribute>.
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: inline code pathmap?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+### `in` method overload resolution tiebreaker
+
+When the `in` argument modifier was added, these two methods would cause an ambiguity:
+
+```csharp
+static void M(S arg);
+static void M(in S arg);
+```
+
+Now, the by value (first in the preceding example) overload is better than the by readonly reference version. To specify the by readonly reference version is called, you must include the `in` modifier when calling the method.
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: the first part of the last sentence is a bit hard to parse. Maybe if you change to something like "To call the by readonly reference version"?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+In earlier versions of C#, you needed to pin a variable to access one of the integers that are part of `myFixedField`. Now, the following code compiles:
+
+```csharp
+var s = new S();
+int p = s.myFixedField[5];
+```
+
+The variable `p` doesn't need to be pinned. Note that you still need an `unsafe` context.
+
+You can learn more in the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: You can learn more -> The boilerplate sentence for this kind of sentence is usually "For more information, see..." or "For more information about x, see ...". I think it helps to localize as well since it's a standard sentence.

(several times on this topic)

File: docs/csharp/whats-new/csharp-7-3.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+Consider this struct:
+
+```csharp
+unsafe struct S
+{
+    public fixed int myFixedField[10];
+}
+```
+
+In earlier versions of C#, you needed to pin a variable to access one of the integers that are part of `myFixedField`. Now, the following code compiles:
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: would it be helpful to show the earlier version code for comparison?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+int p = s.myFixedField[5];
+```
+
+The variable `p` doesn't need to be pinned. Note that you still need an `unsafe` context.
+
+You can learn more in the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
+
+### `ref` local variables may be reassigned
+
+Now, `ref` locals may be reassigned to refer to different storage after being initialized. The following code now compiles:
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: to different storage -> to a different storage?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+```csharp 
+ref VeryLargeStruct reflocal = ref veryLargeStruct; // initialization
+refLocal = ref anotherVeryLargeStruct; // reassigned, refLocal refers to different storage.
+```
+
+You can learn more in the article on [`ref` returns and `ref` locals](../programming-guide/classes-and-structs/ref-returns.md).
+
+### `stackalloc` arrays support initializers
+
+You have been able to specify the values for elements in an array when you initialize the array:
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: nits: You have been -> You've been
initialize the array -> initialize it?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+```csharp
+var arr = new int[3] {1, 2, 3};
+var arr2 = new int[] { 1, 2, 3};
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: nit remove extra space

File: docs/csharp/whats-new/csharp-7-3.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+Now, that same syntax can be applied to arrays that are declared with `stackalloc`:
+
+```csharp
+int* pArr = stackalloc int[3] {1, 2, 3};
+int* pArr2 = stackalloc int[] { 1, 2, 3};
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: nit remove extra space

File: docs/csharp/whats-new/csharp-7-3.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+Now, the by value (first in the preceding example) overload is better than the by readonly reference version. To specify the by readonly reference version is called, you must include the `in` modifier when calling the method.
+
+> [!NOTE]
+> This was implemented as a bug fix. This no longer is ambiguous even with the language version set to "7.2".
+
+For more information, see the article on the [`in` parameter modifier](../language-reference/keywords/in-parameter-modifier.md).
+
+### Extend expression variables in initializers
+
+The syntax added in C# 7.0 to permit out variable declarations has been extended to include field initializers, property initializers, constructor initializers, and query clauses. It enables code such as the following example:
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: replace permit with allow? inline code out?

File: docs/csharp/whats-new/csharp-7-3.md [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+1. When a method group contains some generic methods whose type arguments do not satisfy their constraints, these members are removed from the candidate set.
+1. For a method group conversion, candidate methods whose return type doesn't match up with the delegate's return type are removed from the set.
+
+You'll only notice this change because you'll find fewer compiler errors for ambiguous method overloads when you are sure which method is better.
+
+## New compiler options
+
+New compiler options support new build and DevOps scenarios for C# programs
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: missing punctuation

File: docs/csharp/whats-new/csharp-7-3.md [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+- You can test `==` and `!=` with tuple types.
+- You can use expression variables in more locations.
+- You may attach attributes to the backing field of auto-implemented properties.
+- Method resolution when arguments differ by `in` has been improved.
+- Overload resolution now has fewer ambiguous cases.
+
+The new compiler options are:
+
+- `-publicsign` to enable OSS signing of assemblies.
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-3.md]: spell out OSS on first usage

**Review by mairaw — CHANGES_REQUESTED** [review]
Looks great @BillWagner. Left a couple of comments.

File: docs/csharp/whats-new/csharp-7-3.md (hunk 10 above)
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: I changed to "instances"

File: docs/csharp/whats-new/csharp-7-3.md (hunk 8 above)
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: I reworded all of them.

File: docs/csharp/whats-new/csharp-7-3.md [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+Now, `ref` locals may be reassigned to refer to different storage after being initialized. The following code now compiles:
+
+```csharp 
+ref VeryLargeStruct reflocal = ref veryLargeStruct; // initialization
+refLocal = ref anotherVeryLargeStruct; // reassigned, refLocal refers to different storage.
+```
+
+You can learn more in the article on [`ref` returns and `ref` locals](../programming-guide/classes-and-structs/ref-returns.md).
+
+### `stackalloc` arrays support initializers
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: I'm torn. I hate not styling them, because otherwise the first word should be caps. And the loc issue.

File: docs/csharp/whats-new/csharp-7-3.md [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+
+You can learn more in the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
+
+### Enhanced generic constraints
+
+You can now specify the type <xref:System.Enum?displayProperty=nameWithType> or <xref:System.Delegate?displayProperty=nameWithType> as base class constraints for a type parameter.
+
+You can also use the new `unmanaged` constraint, to specify that a type parameter must be an **unmanaged type**. An **unmanaged type** is a type that isn't a reference type, and doesn't contain any reference type at any level of nesting.
+
+You can learn more in the articles on [`where` generic constraints](../language-reference/keywords/where-generic-type-constraints.md) and the article covering [constraints on type parameters](../programming-guide/generics/constraints-on-type-parameters.md).
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-3.md]: I fixed this one. The others rely on #5408

File: docs/csharp/whats-new/csharp-7-3.md [hunk 19]
```diff
@@ -45,68 +45,90 @@ unsafe struct S
...
+For more information, see the articles on [`where` generic constraints](../language-reference/keywords/where-generic-type-constraint.md) and [constraints on type parameters](../programming-guide/generics/constraints-on-type-parameters.md).
 
 ## Make existing features better
 
 The second theme provides improvements to features in the language. These features improve productivity when writing C#.
 
 ### Tuples support `==` and `!=`
 
-The C# tuple types now support `==` and `!=`. You can learn more about the rules in the article on [tuples](../tuples.md#equality-and-tuples).
+The C# tuple types now support `==` and `!=`. Fore more information, see the section covering [equality](../tuples.md#equality-and-tuples) in the article on [tuples](../tuples.md#equality-and-tuples).
```
Comment (pkulikov) [inline docs/csharp/whats-new/csharp-7-3.md]: nit: the sentence should end with the link to the article itself, not the section:
`...in the article on [tuples](../tuples.md)`

File: docs/csharp/whats-new/csharp-7-3.md [hunk 20]
```diff
@@ -137,7 +159,7 @@ For more information, see the article on the [`in` parameter modifier](../langua
 
 ### Extend expression variables in initializers
 
-The syntax added in C# 7.0 to permit out variable declarations has been extended to include field initializers, property initializers, constructor initializers, and query clauses. It enables code such as the following example:
+The syntax added in C# 7.0 to aloow `out` variable declarations has been extended to include field initializers, property initializers, constructor initializers, and query clauses. It enables code such as the following example:
```
Comment (pkulikov) [inline docs/csharp/whats-new/csharp-7-3.md]: Typo: *aloow* -> *allow*

File: docs/csharp/whats-new/csharp-7-3.md [hunk 21]
```diff
@@ -169,16 +191,16 @@ You'll only notice this change because you'll find fewer compiler errors for amb
...
+### Public or Open Source signing
 
 The `-publicsign` compiler option instructs the compiler to sign the assembly using a public key. The assembly is marked as signed, but the signature is taken from the public key. This option enables you to build signed assemblies from open-source projects using a public key.
 
 For more information, see the [-publicsign compiler option](../language-reference/compiler-options/publicsign-compiler-option.md) article.
 
 ### pathmap
 
-The `-pathmap` compiler option instructs the compiler to replace source paths from the build environment with mapped source paths. The pathmap option controls the source path written by the compiler to PDB files or for the <xref:System.Runtime.CompilerServices.CallerFilePathAttribute>.
+The `-pathmap` compiler option instructs the compiler to replace source paths from the build environment with mapped source paths. The `pathmap` option controls the source path written by the compiler to PDB files or for the <xref:System.Runtime.CompilerServices.CallerFilePathAttribute>.
```
Comment (pkulikov) [inline docs/csharp/whats-new/csharp-7-3.md]: nit: add dash to the second mention of `-pathmap`

**Comment by BillWagner** [comment]
closing and reopening to start a new build.

File: docs/csharp/whats-new/csharp-7-3.md [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+    static S s = new S();
+
+    public void M()
+    {
+        int p = s.myFixedField[5];
+    }
+}
+```
+
+The variable `p` doesn't need to be pinned. Note that you still need an `unsafe` context. In earlier versions of C#, you need to declare a second fixed pointer:
```
Comment (svick) [inline docs/csharp/whats-new/csharp-7-3.md]: I think it's inaccurate to say that `p` doesn't need to be pinned. `p` is just an `int`, it can't be pinned and the following example does not pin `p`.

File: docs/csharp/whats-new/csharp-7-3.md [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+
+```csharp
+class C
+{
+    static S s = new S();
+
+    public void M()
+    {
+        int p = s.myFixedField[5];
+    }
```
Comment (svick) [inline docs/csharp/whats-new/csharp-7-3.md]: As the following text says, this code has be in an `unsafe` context, so I think this example should include `unsafe` somewhere (probably on the method).

File: docs/csharp/whats-new/csharp-7-3.md [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+
+Fore more information, see the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
+
+### `ref` local variables may be reassigned
+
+Now, `ref` locals may be reassigned to refer to different instances after being initialized. The following code now compiles:
+
+```csharp
+ref VeryLargeStruct reflocal = ref veryLargeStruct; // initialization
+refLocal = ref anotherVeryLargeStruct; // reassigned, refLocal refers to different storage.
```
Comment (svick) [inline docs/csharp/whats-new/csharp-7-3.md]: Typo: first line uses `reflocal` (with lowecase `l`), second line `refLocal` (with capital `L`).

**Comment by mairaw** [comment]
There are some build warnings here, @BillWagner. I'll take a look at the rest now.

File: docs/csharp/whats-new/csharp-7-3.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (mairaw) [reply docs/csharp/whats-new/csharp-7-3.md]: Keep the styling. My thought process was more related to UX changes. Like, should we work with @adkinn and co. to come up with some rules for when inline code should not be formatted like in headings and links?

**Review by mairaw — APPROVED** [review]
LGTM. @svick had some new comments you might wanna check and you still have one instance of OSS which is not spelled out. You can use the acronym after you show what it means once.

File: docs/csharp/whats-new/csharp-7-3.md [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+Fore more information, see the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
+
+### `ref` local variables may be reassigned
+
+Now, `ref` locals may be reassigned to refer to different instances after being initialized. The following code now compiles:
+
+```csharp
+ref VeryLargeStruct refLocal = ref veryLargeStruct; // initialization
```
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-3.md]: There is one more form of ref local re-assignment: `foreach ref`. It was missing from the spec, but it is implemented and shipped.

```C#
        foreach (ref var v in new int[0])
        {
        }
```

Comment (pkulikov) [reply docs/csharp/whats-new/csharp-7-3.md]: @jcouv that means we can make such shoot-in-the-leg things now:
```csharp
var _array = new byte[5];
new Random(42).NextBytes(_array);
Span<byte> span = _array;
foreach (ref byte element in span)
{
    element = 4;
    Console.WriteLine(element);
}
Console.WriteLine(string.Join(" ", _array));
Console.ReadLine();
```
That outputs only `4`s. Wow.
